integrations/transportation/tfl.js

```diff
@@ -0,0 +1,173 @@
+  }
+
+  const postBody = {
+    username,
+    password,
+    AppId: '9C9C6B6C-A025-493E-8F39-3A6D57C7ACAB', // TODO what do we do with this?
+  };
+
+  const loginResponse = await fetch(LOGIN_PATH, {
+    method: 'post',
+    headers: {
+      'Content-type': 'application/json',
+    },
+    body: JSON.stringify(postBody),
+  })
+    .then(response => response.json())
+    .catch((e) => {
+      throw new HTTPError(e);
+    });
+
+  if (!loginResponse.SecurityToken) throw new HTTPError('Login failed');
+
+  // console.log(loginResponse);
+
+  const apiTokenResponse = await fetch(`${BASE_PATH}/APITokens`, {
+    method: 'get',
+    headers: {
+      code: loginResponse.SecurityToken,
+      grant_type: 'authorization_code',
+    },
+  })
+    .then(response => response.json())
+    .catch((e) => {
+      throw new HTTPError(e);
+    });
+
+  // console.log(apiTokenResponse);
+
+  return {
+    securityToken: loginResponse.SecurityToken,
+    refreshToken: apiTokenResponse.refresh_token,
+    accessToken: apiTokenResponse.access_token,
+  };
+}
+
+async function getOneWeeksOysterData(accessToken, oysterCardNumber, startMoment, endMoment) {
+  return fetch(`${BASE_PATH}/Cards/Oyster/Journeys?startDate=${startMoment.format(moment.HTML5_FMT.DATE)}&endDate=${endMoment.format(moment.HTML5_FMT.DATE)}`, {
+    method: 'get',
+    headers: {
+      'x-zumo-auth': accessToken,
+      oystercardnumber: oysterCardNumber,
+    },
+  })
+    .catch((e) => {
+      throw new HTTPError(e);
+    });
+}
+
+async function collect(state) {
+  // Get access token using the refresh token
+  const refreshTokenResponse = await fetch(`${BASE_PATH}/APITokens/RefreshToken`, {
+    method: 'get',
+    headers: {
+      refresh_token: state.refreshToken,
+      access_token: state.accessToken,
+    },
+  })
+    .then(response => response.json())
+    .catch((e) => {
+      throw new HTTPError(e);
+    });
+
+  // Always update state with new tokens
+  const newState = {
+    accessToken: refreshTokenResponse.access_token,
+    refreshToken: refreshTokenResponse.refresh_token,
+  };
+
+  // Update oyster cards
+  const oysterCardResponse = await fetch(`${BASE_PATH}/Cards/Oyster`, {
+    method: 'get',
+    headers: {
+      'x-zumo-auth': newState.accessToken,
+    },
+  })
+    .then(response => response.json())
+    .catch((e) => {
+      throw new HTTPError(e);
+    });
+
+  const oysterCards = oysterCardResponse.OysterCards;
+
+  newState.oysterCardNumbers = oysterCards.map(oc => oc.OysterCardNumber);
+
+  const today = moment().startOf('day');
+
+  let startDate = moment().subtract(56, 'days').startOf('day'); // Max data range is 56 days
+
+  if (newState.oysterCardNumbers && newState.oysterCardNumbers.length > 0) {
+    const travelDays = []; // Array of objects for each day travelled, with journeys inside them
+    // First, get most recent data (between startDate and today)
+    while (startDate < today) {
+      /* eslint-disable no-await-in-loop */
```

Comment:
certainly not a requirement as this is perfectly good code ❤️, but there might be an opportunity for concurrency here -- by creating an array of promises in the loop then `Promise.all()`ing them outside, you can avoid the `eslint-disable`.

Example:
https://github.com/tmrowco/tmrowapp-contrib/blob/c7c5b5adb57d5e5e852085b7719fc6cc72409eed/integrations/transportation/uber.js#L79

**This is only good if API returns a result count in the response, otherwise you don't know how many results you need to fetch in which case collecting results sequentially is a requirement**

Again, just a thought 🤓

Reply:
I've added concurrency by using `Promise.all()`